File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    private Map<Integer, List<Integer>> m;
    private boolean[] seen;
    private boolean[] checked;
    
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        this.m = new HashMap<>();
        this.seen = new boolean[numCourses];
        this.checked = new boolean[numCourses];
        
        for (int[] prerequisite : prerequisites) {
            if (!m.containsKey(prerequisite[0])) {
                m.put(prerequisite[0], new ArrayList<>());
            }
            m.get(prerequisite[0]).add(prerequisite[1]);
        }
        
        for (int currCourse = 0; currCourse < numCourses; currCourse++) {
            if (isCyclic(currCourse)) {
                return false;
            }
        }
        
        return true;
    }
    
    private boolean isCyclic(int currCourse) {
        if (checked[currCourse]) {
            return false;
        }
        if (seen[currCourse]) {
            return true;
        }
        
        if (!m.containsKey(currCourse)) return false;
        
        seen[currCourse] = true;
        boolean res = false;
        for (int nextCourse : m.get(currCourse)) {
            res = isCyclic(nextCourse);
            if (res) {
                break;
            }
        }
        
        seen[currCourse] = false;
        
        checked[currCourse] = true;
        return res;
    }
}

File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
private:
    unordered_map<int, vector<int>> m;
    unordered_set<int> seen;
    
    bool isCyclic(int currCourse) {
        if (seen.find(currCourse) != seen.end()) {
            return false;
        }
        if (m[currCourse].empty()) {
            return true;
        }
        
        seen.insert(currCourse);
        bool res = false;
        vector<int> tempList = m[currCourse];
        for (int i = 0; i < tempList.size(); i++) {
            int nextCourse = tempList[i];
            if (!isCyclic(nextCourse)) {
                return false;
            }
        }
        m[currCourse].clear();
        seen.erase(currCourse);
        
        return true;
    }

public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        
        for (int i = 0; i < prerequisites.size(); i++) {
            m[prerequisites[i][0]].push_back(prerequisites[i][1]);
        }
        
        for (int currCourse = 0; currCourse < numCourses; currCourse++) {
            if (!isCyclic(currCourse)) {
                return false;
            }
        }
        
        return true;
    }
};